File: algorithm/cpp/firstround/116.cpp
#include <cstdio>
using namespace std;

// Definition for a Node.
class Node {
public:
    int val;
    Node* left;
    Node* right;
    Node* next;

    Node() : val(0), left(NULL), right(NULL), next(NULL) {}

    Node(int _val) : val(_val), left(NULL), right(NULL), next(NULL) {}

    Node(int _val, Node* _left, Node* _right, Node* _next)
        : val(_val), left(_left), right(_right), next(_next) {}
};

//时间复杂度O(N)，空间复杂度O(1)
class Solution {
public:
    Node* connect(Node* root) {
        if(root==nullptr) return root;

        Node* leftestNode = root;
        while(leftestNode->left!=nullptr)
        {
            Node* head = leftestNode;
            while(head!=nullptr)
            {
                head->left->next = head->right;

                if(head->next!=nullptr)
                {
                    head->right->next = head->next->left;
                }
                head = head->next;
            }
            leftestNode = leftestNode->left;
        }
        return root;
    }
};

int main()
{
    return 0;
}

File: algorithm/go/firstround/116.go
package firstround

type Node struct {
	Val int
	Left *Node
	Right *Node
	Next *Node
}

// 层序遍历,时间复杂度O(N),空间复杂度O(N)
// func connect(root *Node) *Node {
// 	if root == nil {
// 		return root
// 	}

// 	nodes := make([]*Node, 0)
// 	nodes = append(nodes, root)

// 	for len(nodes) > 0 {
// 		sz := len(nodes)
// 		for i := 0; i < sz; i++ {
// 			node := nodes[0]

// 			if node.Left != nil {
// 				nodes = append(nodes, node.Left)
// 			}
// 			if node.Right != nil {
// 				nodes = append(nodes, node.Right)
// 			}
// 			nodes = nodes[1:]
			
// 			// 连接next节点
// 			if i < sz-1 {
// 				node.Next = nodes[0]
// 			}
// 		}
// 	}

// 	return root
// }

// 利用Next指针遍历,时间复杂度O(N),空间复杂度O(1)
func connect(root *Node) *Node {
	if root == nil {
		return root
	}

	// 从每一层的最左侧节点开始遍历
	for leftMost := root; leftMost.Left != nil; leftMost = leftMost.Left {
		// 遍历该层节点,为下一层的节点连接Next指针
		for node := leftMost; node != nil; node = node.Next {
			node.Left.Next = node.Right
			if node.Next != nil {
				node.Right.Next = node.Next.Left
			}
		}
	}

	return root
}